### PR TITLE
Send telemetry events to telemetry plugin

### PR DESCRIPTION
**Pull request by dkwon17**
Signed-off-by: David Kwon <dakwon@redhat.com>

Part of https://github.com/eclipse/che/issues/21661

This PR introduces a new built-in extension:
image

Which sends telemetry events to `http://localhost:${DEVWORKSPACE_TELEMETRY_BACKEND_PORT}`

Compared to [Che-theia telemetry](https://github.com/eclipse/che/issues/15469), the Che-code telemetry:

* does not detect the `git.push` and `git.commit` commands ([like in che-theia](https://github.com/eclipse-che/che-theia/blob/8bef58f2223607312e09d7cca62d6df0c00f0ca6/plugins/telemetry-plugin/src/telemetry-plugin.ts#L17-L23)) since `onDidExecuteCommand` does not exist in the VS Code api: https://github.com/microsoft/vscode/issues/78091#issuecomment-518545491. Therefore the `Push to remote in Che`and `Commit locally in Che` telemetry events are not being detected/sent.
* does not detect screen dimensions since the `window` object is not available

### To test this PR:

1. Start a workspace with this URL
```
{CHE_HOST}/#https://github.com/dkwon17/che-code/tree/telemetryService
```

2. After the workspace starts, add the following telemetry plugin under the DevWorkspace's `spec.template.components`:
```
- attributes:
  name: telemetry
  plugin:
    uri: https://gist.githubusercontent.com/dkwon17/d2f800494959dadd1106512b0c56275e/raw/5fada3206ee6dba868fd73086acc21d46682cce8/woopra-test.yaml
```
This telemetry plugin will log the received events.
3. Start the workspace again
4. Run `yarn && yarn watch` and then run `yarn server` to install dependencies and start up a new instance of VS Code.
5. Click on `Open In New Tab` to open the new instance
image
6. View the container logs for the telemetry plugin. You could try to send and verify that the `Start Workspace in Che`, `Open Workspace in Che`, and `Edit Workspace File in Che` events are being sent:

`Start Workspace in Che` log (when opening the workspace)
```
2022-10-03 21:40:14,489 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Start Workspace in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723}
```

`Refresh Workspace in Che` log (you should see this when you refresh the page with F5 for example)
```
2022-10-03 21:51:09,529 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Refresh Workspace in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723}
```

`Edit Workspace File in Che` log (when editing a file)
```
2022-10-03 21:40:50,852 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Edit Workspace File in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723, programming language=json}
```

There are more [telemetry events](https://github.com/che-incubator/che-workspace-telemetry-client/blob/4a6e6a8eae0be050a370823c6970faf182c5365e/backend-base/src/main/java/org/eclipse/che/incubator/workspace/telemetry/base/AnalyticsEvent.java#L17-L24), but the `Keep Workspace Inactive in Che` and `Stop Workspace in Che` events are handled form the plugin side and not the VS Code extension side (See [1](https://github.com/che-incubator/devworkspace-telemetry-woopra-plugin/blob/34bf91ed407ca49ceea1256c20599f72137ddee0/src/main/java/com/redhat/devworkspace/services/telemetry/woopra/AnalyticsManager.java#L188-L192), [2](https://github.com/che-incubator/devworkspace-telemetry-woopra-plugin/blob/34bf91ed407ca49ceea1256c20599f72137ddee0/src/main/java/com/redhat/devworkspace/services/telemetry/woopra/AnalyticsManager.java#L431-L432)).

### CODE REVIEW

File: code/build/gulpfile.extensions.js [hunk 1]
```diff
@@ -31,12 +31,14 @@ const extensionsPath = path.join(path.dirname(__dirname), 'extensions');
 // });
 const compilations = [
 	'authentication-proxy/tsconfig.json',
+	'che-activity-tracker/tsconfig.json',
```
Comment (azatsarynnyy) [inline code/build/gulpfile.extensions.js:34]: So, the idled workspaces are not stopped... 
Good catch! :+1:

File: code/extensions/che-telemetry/extension.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/**********************************************************************
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/extension.ts]: This is redundant. Right?
The extension entrypoint is placed in `src`.

File: code/extensions/che-telemetry/package.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        "@types/fs-extra": "^9.0.13",
+        "@types/jest": "^27.4.1",
+        "@types/js-yaml": "^4.0.5",
+        "@types/node": "14.x",
+        "add": "^2.0.6",
+        "jest": "27.3.1",
+        "ts-jest": "^27.1.4",
+        "yarn": "^1.22.18"
+    },
+    "dependencies": {
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/package.json]: Both dependencies aren't used in the code.

File: code/extensions/che-telemetry/package.json [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    },
+    "main": "./out/extension.js",
+    "scripts": {
+        "compile": "gulp compile-extension:che-telemetry",
+        "watch": "gulp watch-extension:che-telemetry",
+        "vscode:prepublish": "npm run compile",
+        "test": "jest",
+        "lint:fix": "eslint --fix --cache=true --no-error-on-unmatched-pattern=true \"{src,tests}/**/*.{ts,tsx}\""
+    },
+    "devDependencies": {
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/package.json]: It looks like none of the declared devDependencies are actually being used.

File: code/extensions/che-api/package.json [hunk 5]
```diff
@@ -36,7 +36,8 @@
     "inversify": "^6.0.1",
     "js-yaml": "^4.1.0",
     "reflect-metadata": "^0.1.13",
-    "vscode-nls": "^5.0.0"
+    "vscode-nls": "^5.0.0",
+    "@eclipse-che/workspace-telemetry-client": "latest"
```
Comment (azatsarynnyy) [inline code/extensions/che-api/package.json:40]: Let's use tag `^0.0.1-1654006444` to keep the build reproducible.

File: code/extensions/che-telemetry/package.nls.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
+	"displayName": "Eclipse Che Telemetry Extension",
+	"description": "Eclipse Che Telemetry Extension"
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/package.nls.json]: Let's use something more descriptive for the description, e.g. `Detects and sends the events to a backend telemetry plugin`.

File: code/extensions/che-telemetry/src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+import * as vscode from 'vscode';
+import { TelemetryEventService } from "./telemetry-event-service";
+
+let telemetryEventService: TelemetryEventService;
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/src/extension.ts]: Is there a reason to move the `telemetryEventService` outside of the `activate` function?
Can we declare `telemetryEventService` as a constant within the `activate`?

File: code/extensions/che-telemetry/telemetry-event-service.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
+/**********************************************************************
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/telemetry-event-service.ts]: a dup of the `src/telemetry-event-service.ts`?

File: code/extensions/che-telemetry/constants.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+/**********************************************************************
```
Comment (azatsarynnyy) [inline code/extensions/che-telemetry/constants.ts]: a dup of `src/constants.ts`?

**Review by azatsarynnyy — COMMENTED** [review]
It seems the `code/extensions/che-telemetry/telemetry-0.0.1.vsix` build artefact is not need there.

File: code/extensions/che-api/package.json (hunk 5 above)
Comment (azatsarynnyy) [inline code/extensions/che-api/package.json]: Let's use tag `^0.0.1-1654006444`.

File: code/extensions/che-telemetry/extension.ts (hunk 2 above)
Comment (dkwon17) [reply code/extensions/che-telemetry/extension.ts]: Oops, yes, I forgot to clean up 🤦

File: code/extensions/che-telemetry/package.nls.json (hunk 6 above)
Comment (dkwon17) [reply code/extensions/che-telemetry/package.nls.json]: Updated to:
```
Detects and sends telemetry events to a backend telemetry plugin via the Eclipse Che API built-in extension
```

File: code/extensions/che-telemetry/src/extension.ts (hunk 7 above)
Comment (dkwon17) [reply code/extensions/che-telemetry/src/extension.ts]: Fixed, moved to be inside the `activate` function

**Comment by dkwon17** [comment]
Thank you for the review @azatsarynnyy , I removed the duplicate files, updated the dependencies & extension description, and addressed the other feedback

**Comment by azatsarynnyy** [comment]
@dkwon17 there are still a couple of redundant files: 
- `code/extensions/che-telemetry/telemetry-0.0.1.vsix`
- `code/extensions/che-telemetry/telemetry-event-service.ts`

**Comment by azatsarynnyy** [comment]
I tried to test it following the provided steps. I added the telemetry plugin to DevWorkspace.

<details>
  <summary>Here's how it looks after adding the plugin.</summary>

```yaml
apiVersion: workspace.devfile.io/v1alpha1
kind: DevWorkspace
metadata:
  annotations:
    controller.devfile.io/timing.deployment.ready: '1665137399262'
    controller.devfile.io/timing.deployment.created: '1665137389988'
    che.eclipse.org/last-updated-timestamp: '2022-10-07T10:09:49.785Z'
    controller.devfile.io/timing.routing.ready: '1665137389986'
    controller.devfile.io/timing.components.ready: '1665137389922'
    controller.devfile.io/timing.routing.created: '1665137389923'
    controller.devfile.io/timing.started: '1665137389902'
    che.eclipse.org/devfile-source: |
      scm:
        repo: 'https://github.com/dkwon17/che-code.git'
        revision: telemetryService
        fileName: devfile.yaml
      factory:
        params: 'url=https://github.com/dkwon17/che-code/tree/telemetryService'
    controller.devfile.io/timing.components.created: '1665137389902'
    che.eclipse.org/che-editor: che-incubator/che-code/insiders
  resourceVersion: '929547384'
  name: che-code
  uid: 528ff176-15e9-4208-891c-6fd120a3bd9e
  creationTimestamp: '2022-10-07T09:44:34Z'
  generation: 8
  managedFields:
    - apiVersion: workspace.devfile.io/v1alpha2
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:annotations':
            .: {}
            'f:che.eclipse.org/che-editor': {}
            'f:che.eclipse.org/devfile-source': {}
            'f:che.eclipse.org/last-updated-timestamp': {}
        'f:spec':
          .: {}
          'f:routingClass': {}
          'f:started': {}
          'f:template':
            .: {}
            'f:attributes':
              .: {}
              'f:controller.devfile.io/devworkspace-config':
                .: {}
                'f:name': {}
                'f:namespace': {}
              'f:controller.devfile.io/storage-type': {}
              'f:dw.metadata.annotations':
                .: {}
                'f:che.eclipse.org/che-editor': {}
                'f:che.eclipse.org/devfile-source': {}
                'f:che.eclipse.org/last-updated-timestamp': {}
            'f:commands': {}
            'f:projects': {}
      manager: unknown
      operation: Update
      time: '2022-10-07T09:44:37Z'
    - apiVersion: workspace.devfile.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:spec':
          'f:template':
            'f:components': {}
      manager: Mozilla
      operation: Update
      time: '2022-10-07T10:05:45Z'
    - apiVersion: workspace.devfile.io/v1alpha2
      fieldsType: FieldsV1
      fieldsV1:
        'f:status':
          'f:conditions': {}
          'f:message': {}
      manager: devworkspace-controller
      operation: Update
      time: '2022-10-07T10:10:01Z'
  namespace: artem-che
  finalizers:
    - storage.controller.devfile.io
  labels:
    controller.devfile.io/creator: 42925d7d-de9b-4bec-853b-f9a607279489
spec:
  routingClass: che
  started: true
  template:
    commands:
      - exec:
          commandLine: |
            yarn prepare
          component: dev
          group:
            kind: build
          id: prepare
          workingDir: '${PROJECTS_ROOT}/che-code'
      - exec:
          commandLine: |
            yarn watch
          component: dev
          group:
            isDefault: true
            kind: build
          id: build
          workingDir: '${PROJECTS_ROOT}/che-code'
      - exec:
          commandLine: |
            yarn server
          component: dev
          group:
            isDefault: true
            kind: run
          id: run
          workingDir: '${PROJECTS_ROOT}/che-code'
    components:
      - plugin:
          name: telemetry
          uri: >-
            https://gist.githubusercontent.com/dkwon17/d2f800494959dadd1106512b0c56275e/raw/5fada3206ee6dba868fd73086acc21d46682cce8/woopra-test.yaml
      - container:
          command:
            - /checode/entrypoint-volume.sh
          endpoints:
            - exposure: public
              name: dev
              protocol: http
              secure: true
              targetPort: 8000
            - attributes:
                contributed-by: che-code.eclipse.org
                cookiesAuthEnabled: ''
                discoverable: ''
                type: main
                urlRewriteSupported: ''
              exposure: public
              name: che-code
              path: '?tkn=eclipse-che'
              protocol: https
              targetPort: 3100
            - attributes:
                contributed-by: che-code.eclipse.org
                discoverable: ''
                urlRewriteSupported: ''
              exposure: public
              name: code-redirect-1
              protocol: http
              targetPort: 13131
            - attributes:
                contributed-by: che-code.eclipse.org
                discoverable: ''
                urlRewriteSupported: ''
              exposure: public
              name: code-redirect-2
              protocol: http
              targetPort: 13132
            - attributes:
                contributed-by: che-code.eclipse.org
                discoverable: ''
                urlRewriteSupported: ''
              exposure: public
              name: code-redirect-3
              protocol: http
              targetPort: 13133
          env:
            - name: CHE_DASHBOARD_URL
              value: 'https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com'
            - name: CHE_PLUGIN_REGISTRY_URL
              value: >-
                https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com/plugin-registry/v3
            - name: CHE_PLUGIN_REGISTRY_INTERNAL_URL
              value: 'http://plugin-registry.dogfooding.svc:8080/v3'
            - name: OPENVSX_REGISTRY_URL
              value: 'https://open-vsx.org'
          image: 'quay.io/che-incubator/che-code-dev:insiders'
          memoryLimit: 8Gi
          name: dev
          sourceMapping: /projects
          volumeMounts:
            - name: checode
              path: /checode
      - volume:
          name: projects
          size: 3Gi
      - plugin:
          kubernetes:
            name: che-code-workspace528ff17615e94208
            namespace: artem-che
          name: che-code-workspace528ff17615e94208
    projects:
      - git:
          checkoutFrom:
            revision: telemetryService
          remotes:
            origin: 'https://github.com/dkwon17/che-code.git'
        name: che-code
status:
  conditions:
    - lastTransitionTime: '2022-10-07T10:09:49Z'
      message: DevWorkspace is starting
      status: 'True'
      type: Started
    - lastTransitionTime: '2022-10-07T10:09:49Z'
      message: Resolved plugins and parents from DevWorkspace
      status: 'True'
      type: DevWorkspaceResolved
    - lastTransitionTime: '2022-10-07T10:09:49Z'
      message: Storage ready
      status: 'True'
      type: StorageReady
    - lastTransitionTime: '2022-10-07T10:09:50Z'
      message: Networking ready
      status: 'True'
      type: RoutingReady
    - lastTransitionTime: '2022-10-07T10:09:50Z'
      message: DevWorkspace serviceaccount ready
      status: 'True'
      type: ServiceAccountReady
    - lastTransitionTime: '2022-10-07T10:09:50Z'
      message: DevWorkspace secrets ready
      status: 'True'
      type: PullSecretsReady
    - lastTransitionTime: '2022-10-07T10:10:01Z'
      message: DevWorkspace deployment ready
      status: 'True'
      type: DeploymentReady
    - lastTransitionTime: '2022-10-07T10:10:01Z'
      message: Waiting for editor to start
      status: 'False'
      type: Ready
  ideUrl: >-
    https://workspace528ff17615e94208-2.apps.che-dev.x6e0.p1.openshiftapps.com/?tkn=eclipse-che
  message: Waiting for editor to start
  phase: Starting
  workspaceId: workspace528ff17615e94208
```
</details>

After restarting the workspace, I got:
image

The logs from Che Dashboard:
```
Initializing DevWorkspace
Preparing networking
Waiting for workspace deployment
Waiting for editor to start
devworkspace failed to progress past phase 'Starting' for longer than timeout (5m)
Initializing DevWorkspace
Preparing networking
Waiting for workspace deployment
Waiting for editor to start
```

From the containers logs, I see that all containers are started.
Am I doing something wrong?

Note, before adding the telemetry plugin to DevWorkspace, it starts successfully.

**Review by azatsarynnyy — COMMENTED** [review]
I was able to test the changes in a bit different way.
I prepared a test repo https://github.com/azatsarynnyy/che-code/tree/telemetry-3, where I declared the custom editor (based on this PR) and added the telemetry plugin definition directly to the Devfile: https://github.com/dkwon17/che-code/compare/main...azatsarynnyy:che-code:telemetry-3?expand=1

After starting a DevWorkspace, I checked the logs in the telemetry plugin's container. There are expected messages:

```
2022-10-07 11:34:45,947 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-1) sending Refresh Workspace in Che (ip= - userAgent=) with properties: {creation time=1665140898000, devworkspace id=workspaceabe1e31c995344a6, devworkspace name=che-code, start time=1665140900999, age=
...
2022-10-07 11:35:47,083 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-1) sending Edit Workspace File in Che (ip= - userAgent=) with properties: {creation time=1665140898000, devworkspace id=workspaceabe1e31c995344a6, devworkspace name=che-code, start time=1665140900999, age=2, programming language=markdown}
```

**Comment by dkwon17** [comment]
> @dkwon17 there are still a couple of redundant files:

Sorry, fixed

> I tried to test it following the provided steps. I added the telemetry plugin to DevWorkspace.

I was able to get your sample working by first converting it from v1alpha1 to a v1alpha2 DevWorkspace by running:
```
oc get devworkspace.v1alpha2.workspace.devfile.io che-code-test-4 -o yaml > output.yaml
```

and then in `output.yaml`'s `dev` container's endpoints' attributes, I set:
```
cookiesAuthEnabled: true     (if exists on endpoint)
discoverable: false
urlRewriteSupported: true
```
since I noticed they were empty strings in the v1alpha1 DevWorkspace. I'm not sure why they were empty in the first place, maybe there is a bug? I haven't seen this issue when working on this PR

> After starting a DevWorkspace, I checked the logs in the telemetry plugin's container. There are expected messages:

Looks good, thank you for checking

**Review by azatsarynnyy — APPROVED** [review]
The code looks good and it works as expected.
Thanks, @dkwon17!

**Pull request merged**